src/sensor/ping.cpp

```diff
@@ -359,25 +359,33 @@ void Ping::flash(const QString& fileUrl, bool sendPingGotoBootloader, int baud,
 
     qCDebug(PING_PROTOCOL_PING) << "Finish connection.";
     // TODO: Move thread delay to something more.. correct.
-    QThread::msleep(1000);
-    link()->finishConnection();
+    auto flashSensor = std::function<void(void)>();
+    auto finishConnection = std::function<void(void)>();
+
+    flashSensor = [=] {
+        flasher()->setBaudRate(baud);
+        flasher()->setFirmwarePath(fileUrl);
+        flasher()->setLink(link()->configuration()[0]);
+        flasher()->setVerify(verify);
+        flasher()->flash();
+    };
+
+    finishConnection = [=] {
+        QThread::msleep(1000);
+        link()->finishConnection();
 
-    QSerialPortInfo pInfo(serialLink->port()->portName());
-    QString portLocation = pInfo.systemLocation();
+        QSerialPortInfo pInfo(serialLink->port()->portName());
+        QString portLocation = pInfo.systemLocation();
 
-    qCDebug(PING_PROTOCOL_PING) << "Save sensor configuration.";
-    updatePingConfigurationSettings();
+        qCDebug(PING_PROTOCOL_PING) << "Save sensor configuration.";
+        updatePingConfigurationSettings();
 
-    qCDebug(PING_PROTOCOL_PING) << "Start flash.";
-    QThread::msleep(1000);
+        qCDebug(PING_PROTOCOL_PING) << "Start flash.";
 
-    flasher()->setBaudRate(baud);
-    flasher()->setFirmwarePath(fileUrl);
-    flasher()->setLink(link()->configuration()[0]);
-    flasher()->setVerify(verify);
-    flasher()->flash();
+        QTimer::singleShot(1000, flashSensor);
+    };
+    QTimer::singleShot(1000, finishConnection);
 
-    QThread::msleep(500);
```

Comment:
Is there anything wrong with sleeping the main thread? Can you explain what the hiccups in the interface is?

Reply:
you should never sleep the main thread :) it hangs the QEventLoop